Teuchos:  Clang 10 compiler warning

## Bug Report
@trilinos/teuchos
@shardest

### Description
Using clang-10 to build Salinas from the master Sierra branch, we are getting 3095 compiler warnings related to the file Teuchos_as.hpp. This is against whatever Trilinos is currently in Sierra – not Trilinos integration, although it appears that this file has not changed in some time. There are several hundred more warnings that I have not examined; they may or may not have anything to do with Trilinos. The following is a typical example (starting at Teuchos_as.hpp:1392):

```c++
  static unsigned long long safeConvert (const double t) {
    const unsigned long long minVal = 0; // unsigned, so min value is 0.                          
    const unsigned long long maxVal = std::numeric_limits<unsigned long long>::max ();

    TEUCHOS_TEST_FOR_EXCEPTION(
      t < minVal || t > maxVal,
      std::range_error,
      "Teuchos::ValueTypeConversionTraits<unsigned long long, double>::safeConvert: "
      "Input double t = " << t << " is out of the valid range [" << minVal
      << ", " << maxVal << "] for conversion to unsigned long long.");

    return static_cast<unsigned long long> (t);
  }
```

```
Teuchos_as.hpp:1396:55: warning: implicit conversion from 'const unsigned long long' to 'double' changes value from 18446744073709551615 to 18446744073709551616 [-Wimplicit-int-float-conversion]
```

It appears to be implicitly casting maxVal to a double for the comparison against t, which is bad. I would think we’d want both values to be explicitly the type with larger range before doing a comparison. There are many related functions in this file with the same issue. I’m surprised that there isn’t yet some kind of STL function that does this?

Adding additional information from @shardest in an email chain:
-----------------
You can also simply compile the attached program with clang-10 – I think it is a good model of both the warning behavior and the original program:

clang++ -o warn_me warn_me.cpp 

warn_me.cpp:8:27: warning: implicit conversion from 'const unsigned long long' to 'double' changes value from 18446744073709551615 to 18446744073709551616 [-Wimplicit-int-float-conversion]
    if (t < minVal || t > maxVal) {
                        ~ ^~~~~~
1 warning generated.

The program is adapted from the function referenced in my previous e-mail. It looks like the intent is to check for overflow – not roundoff, which is to be expected. The first integer that can’t be represented by double-precision float is 2^53+1. By the time we get to doubles in the vicinity of 2^64, we are missing a bunch of digits on the end, so it’s a little hazy what should happen. However, I expected this program as written to either throw or return a large integer. In fact, it returns zero because:

•	2^64-1 becomes 2^64 when cast to double
•	2^64 is exactly representable as double

Thus, t == maxVal, so it doesn’t throw, but the conversion overflows. I didn’t make any effort to enable floating point exceptions, which I presume would have caught this, but it seems that the point of the function called “safeConvert” is to preclude this possibility.

    Sean


-------warn_me.cpp-------------
#include <limits>
#include <iostream>

unsigned long long safeConvert (const double t) {
    const unsigned long long minVal = 0; // unsigned, so min value is 0.
    const unsigned long long maxVal = std::numeric_limits<unsigned long long>::max ();
 
    if (t < minVal || t > maxVal) {
        throw std::range_error("double to long long conversion out of bounds");
    }
 
    return static_cast<unsigned long long> (t);
}

int main(void) {
    // 2^53 = 9007199254740992 = 9.007199254740992e15
    // 2^64 = 18446744073709551616 = 1.8446744073709551616e19

    // 2^53 is still representable as double
    std::cout << safeConvert(9007199254740992.0) << std::endl;

    // 2^53 + 1 is not
    std::cout << safeConvert(9007199254740993.0) << std::endl;

    // 2^64
    std::cout << safeConvert(1.8446744073709551616e19) << std::endl;

    return 0;
}

--- a/packages/teuchos/core/src/Teuchos_ScalarTraits.hpp
+++ b/packages/teuchos/core/src/Teuchos_ScalarTraits.hpp
@@ -649,7 +649,7 @@ struct ScalarTraits<float>
     random();
 #endif
   }
-  static inline float random() { float rnd = (float) std::rand() / RAND_MAX; return (-1.0f + 2.0f * rnd); }
+  static inline float random() { float rnd = (float) std::rand() / static_cast<float>(RAND_MAX); return (-1.0f + 2.0f * rnd); }
   static inline std::string name() { return "float"; }
   static inline float squareroot(float x)
     {

--- a/packages/teuchos/core/src/Teuchos_as.hpp
+++ b/packages/teuchos/core/src/Teuchos_as.hpp
@@ -1339,7 +1339,7 @@ public:
     const unsigned long maxVal = std::numeric_limits<unsigned long>::max ();
 
     TEUCHOS_TEST_FOR_EXCEPTION(
-      t < minVal || t > maxVal,
+      t < minVal || t > static_cast<double>(maxVal),
       std::range_error,
       "Teuchos::ValueTypeConversionTraits<unsigned long, double>::safeConvert: "
       "Input double t = " << t << " is out of the valid range [" << minVal
@@ -1394,7 +1394,7 @@ public:
     const unsigned long long maxVal = std::numeric_limits<unsigned long long>::max ();
 
     TEUCHOS_TEST_FOR_EXCEPTION(
-      t < minVal || t > maxVal,
+      t < minVal || t > static_cast<double>(maxVal),
       std::range_error,
       "Teuchos::ValueTypeConversionTraits<unsigned long long, double>::safeConvert: "
       "Input double t = " << t << " is out of the valid range [" << minVal
@@ -1544,7 +1544,7 @@ public:
     const unsigned int maxVal = std::numeric_limits<unsigned int>::max ();
 
     TEUCHOS_TEST_FOR_EXCEPTION(
-      t < minVal || t > maxVal,
+      t < minVal || t > static_cast<float>(maxVal),
       std::range_error,
       "Teuchos::ValueTypeConversionTraits<unsigned int, float>::safeConvert: "
       "Input double t = " << t << " is out of the valid range [" << minVal
@@ -1593,7 +1593,7 @@ public:
     // GNU/Linux and other operating systems).
     if (sizeof (long) < sizeof (float)) {
       TEUCHOS_TEST_FOR_EXCEPTION(
-        t < minVal || t > maxVal,
+        t < minVal || t > static_cast<float>(maxVal),
         std::range_error,
         "Teuchos::ValueTypeConversionTraits<long, float>::safeConvert: "
         "Input float t = " << t << " is out of the valid range ["
@@ -1621,7 +1621,7 @@ public:
     const unsigned long maxVal = std::numeric_limits<unsigned long>::max ();
 
     TEUCHOS_TEST_FOR_EXCEPTION(
-      t < minVal || t > maxVal,
+      t < minVal || t > static_cast<float>(maxVal),
       std::range_error,
       "Teuchos::ValueTypeConversionTraits<unsigned long, float>::safeConvert: "
       << "Input float t = " << t << " is out of the valid range [" << minVal
@@ -1664,7 +1664,7 @@ public:
     const unsigned long long maxVal = std::numeric_limits<unsigned long long>::max ();
 
     TEUCHOS_TEST_FOR_EXCEPTION(
-      t < minVal || t > maxVal,
+      t < minVal || t > static_cast<float>(maxVal),
       std::range_error,
       "Teuchos::ValueTypeConversionTraits<unsigned long long, float>::safeConvert: "
       "Input float t = " << t << " is out of the valid range [" << minVal

--- a/packages/teuchos/core/test/TypeConversions/TypeConversions_UnitTest.cpp
+++ b/packages/teuchos/core/test/TypeConversions/TypeConversions_UnitTest.cpp
@@ -568,7 +568,7 @@ TEUCHOS_UNIT_TEST_TEMPLATE_2_DECL( asSafe, realToUnsignedIntTypeOverflow, RealTy
   TEST_THROW(val = asSafe<UnsignedIntType> (minusOne), std::range_error);
 
   // Only test overflow checks if overflow can actually take place.
-  if (maxUnsignedIntVal < maxVal) {
+  if (static_cast<RealType>(maxUnsignedIntVal) < maxVal) {
     TEST_THROW(val = asSafe<UnsignedIntType> (maxVal), std::range_error);
     try {
       std::cerr << std::endl

